core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -47,55 +47,83 @@ private[server] class MetadataCache(brokerId: Int) extends Logging {
-                  throw new ReplicaNotAvailableException("Replica information not available for following brokers: " +
-                    replicas.filterNot(replicaInfo.map(_.id).contains(_)).mkString(","))
-                if (isrInfo.size < isr.size)
-                  throw new ReplicaNotAvailableException("In Sync Replica information not available for following brokers: " +
-                    isr.filterNot(isrInfo.map(_.id).contains(_)).mkString(","))
-                new PartitionMetadata(partitionId, leaderInfo, replicaInfo, isrInfo, Errors.NONE.code)
-              } catch {
-                case e: Throwable =>
-                  debug("Error while fetching metadata for %s: %s".format(topicPartition, e.toString))
-                  new PartitionMetadata(partitionId, leaderInfo, replicaInfo, isrInfo,
-                    Errors.forException(e).code)
-              }
-          }
-          topicResponses += new TopicMetadata(topic, partitionMetadata.toSeq)
+  private def getPartitionMetadata(topic: String, protocol: SecurityProtocol): List[MetadataResponse.PartitionMetadata] = {
+    cache(topic).map { case (partitionId, partitionState) =>
+      val topicPartition = TopicAndPartition(topic, partitionId)
+
+      val leaderAndIsr = partitionState.leaderIsrAndControllerEpoch.leaderAndIsr
+      val leader = aliveBrokers.get(leaderAndIsr.leader).map(_.getNode(protocol))
+
+      val replicas = partitionState.allReplicas
+      val replicaInfo = getAliveEndpoints(replicas, protocol)
+
+      if (leader.isEmpty) {
+        debug("Error while fetching metadata for %s: leader not available".format(topicPartition))
+        new MetadataResponse.PartitionMetadata(Errors.LEADER_NOT_AVAILABLE, partitionId, Node.noNode(),
+          replicaInfo.asJava, java.util.Collections.emptyList())
+      } else {
+        val isr = leaderAndIsr.isr
+        val isrInfo = getAliveEndpoints(isr, protocol)
+
+        if (replicaInfo.size < replicas.size) {
+          debug("Error while fetching metadata for %s: replica information not available for following brokers %s"
+            .format(topicPartition, replicas.filterNot(replicaInfo.map(_.id).contains(_)).mkString(",")))
+
+          new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+            replicaInfo.asJava, isrInfo.asJava)
+        } else if (isrInfo.size < isr.size) {
+          debug("Error while fetching metadata for %s: in sync replica information not available for following brokers %s"
+            .format(topicPartition, isr.filterNot(isrInfo.map(_.id).contains(_)).mkString(",")))
+          new MetadataResponse.PartitionMetadata(Errors.REPLICA_NOT_AVAILABLE, partitionId, leader.get,
+            replicaInfo.asJava, isrInfo.asJava)
+        } else {
+          new MetadataResponse.PartitionMetadata(Errors.NONE, partitionId, leader.get, replicaInfo.asJava,
+            isrInfo.asJava)
         }
       }
+    }.toList
```

Comment:
Sorry, we don't convert it, but we call `asJava` on it. `scala.List` is an immutable linked list with O(n) size which has different performance characteristics than Java's `List` implementations. It seems like we probably want to call `toBuffer` or something like that (`ArrayBuffer` is similar to `ArrayList` is the most used `java.util.List` implementation).